processor/batchprocessor/batch_processor.go

```diff
@@ -98,120 +199,188 @@ func (bp *batchProcessor) Capabilities() consumer.Capabilities {
-			if bp.batch.itemCount() > 0 {
+			if b.batch.itemCount() > 0 {
 				// TODO: Set a timeout on sendTraces or
 				// make it cancellable using the context that Shutdown gets as a parameter
-				bp.sendItems(triggerTimeout)
+				b.sendItems(triggerTimeout)
 			}
 			return
-		case item := <-bp.newItem:
+		case item := <-b.newItem:
 			if item == nil {
 				continue
 			}
-			bp.processItem(item)
-		case <-bp.timer.C:
-			if bp.batch.itemCount() > 0 {
-				bp.sendItems(triggerTimeout)
+			b.processItem(item)
+		case <-b.timer.C:
+			if b.batch.itemCount() > 0 {
+				b.sendItems(triggerTimeout)
 			}
-			bp.resetTimer()
+			b.resetTimer()
 		}
 	}
 }
 
-func (bp *batchProcessor) processItem(item any) {
-	bp.batch.add(item)
+func (b *batcher) processItem(item any) {
+	b.batch.add(item)
 	sent := false
-	for bp.batch.itemCount() >= bp.sendBatchSize {
+	for b.batch.itemCount() >= b.processor.sendBatchSize {
 		sent = true
-		bp.sendItems(triggerBatchSize)
+		b.sendItems(triggerBatchSize)
 	}
 
 	if sent {
-		bp.stopTimer()
-		bp.resetTimer()
+		b.stopTimer()
+		b.resetTimer()
 	}
 }
 
-func (bp *batchProcessor) stopTimer() {
-	if !bp.timer.Stop() {
-		<-bp.timer.C
+func (b *batcher) stopTimer() {
+	if !b.timer.Stop() {
+		<-b.timer.C
 	}
 }
 
-func (bp *batchProcessor) resetTimer() {
-	bp.timer.Reset(bp.timeout)
+func (b *batcher) resetTimer() {
+	b.timer.Reset(b.processor.timeout)
 }
 
-func (bp *batchProcessor) sendItems(trigger trigger) {
-	sent, bytes, err := bp.batch.export(bp.exportCtx, bp.sendBatchMaxSize, bp.telemetry.detailed)
+func (b *batcher) sendItems(trigger trigger) {
+	sent, bytes, err := b.batch.export(b.exportCtx, b.processor.sendBatchMaxSize, b.processor.telemetry.detailed)
 	if err != nil {
-		bp.logger.Warn("Sender failed", zap.Error(err))
+		b.processor.logger.Warn("Sender failed", zap.Error(err))
 	} else {
-		bp.telemetry.record(trigger, int64(sent), int64(bytes))
+		b.processor.telemetry.record(trigger, int64(sent), int64(bytes))
+	}
+}
+
+func (bp *batchProcessor) findBatcher(ctx context.Context) (*batcher, error) {
+	if bp.singleton != nil {
+		return bp.singleton, nil
+	}
+
+	// Get each metadata key value, form the corresponding
+	// attribute set for use as a map lookup key.
+	info := client.FromContext(ctx)
+	md := map[string][]string{}
+	var attrs []attributeKeyValue
+	for _, k := range bp.metadataKeys {
+		// Lookup the value in the incoming metadata, copy it
+		// into the outgoing metadata, and create a unique
+		// value for the attributeSet.
+		vs := info.Metadata.Get(k)
+		md[k] = vs
+		if len(vs) == 1 {
+			attrs = append(attrs, attributeString(k, vs[0]))
+		} else {
+			attrs = append(attrs, attributeStringSlice(k, vs))
+		}
+	}
+	aset := newAttributeSet(attrs...)
+
+	bp.lock.Lock()
+	defer bp.lock.Unlock()
+
+	b, ok := bp.batchers[aset]
+	if ok {
+		return b, nil
+	}
+
+	if limit := bp.metadataLimit; limit != 0 && len(bp.batchers) >= int(limit) {
+		return nil, errTooManyBatchers
```

Comment:
What do we do with batchers that don't receive any data for a while? Shouldn't we add some expiration mechanism for them?

Reply:
Adding an expiration mechanism is not out of the question, but it raises new synchronization issues that I wanted to bypass. I do not expect that expiration would matter in the cases that we care about.